File: demoapps/starwars-java/src/main/java/com/example/starwars/service/viaduct/ViaductConfiguration.java
package com.example.starwars.service.viaduct;

import io.micronaut.context.annotation.Bean;
import io.micronaut.context.annotation.Factory;
import java.util.List;
import java.util.Set;
import viaduct.service.BasicViaductFactory;
import viaduct.service.SchemaScopeInfo;
import viaduct.service.api.Viaduct;

/** Registers the public and extras schema variants. */
@Factory
public final class ViaductConfiguration {
  public static final String DEFAULT_SCOPE_ID = "default";
  public static final String EXTRAS_SCOPE_ID = "extras";
  public static final SchemaScopeInfo.Scoped DEFAULT_SCHEMA =
      new SchemaScopeInfo.Scoped("publicSchema", Set.of(DEFAULT_SCOPE_ID));
  public static final SchemaScopeInfo.Scoped EXTRAS_SCHEMA =
      new SchemaScopeInfo.Scoped(
          "publicSchemaWithExtras", Set.of(DEFAULT_SCOPE_ID, EXTRAS_SCOPE_ID));

  private final MicronautTenantModuleInjectorFactory tenantModuleInjectorFactory;

  public ViaductConfiguration(MicronautTenantModuleInjectorFactory tenantModuleInjectorFactory) {
    this.tenantModuleInjectorFactory = tenantModuleInjectorFactory;
  }

  @Bean
  public Viaduct providesViaduct() {
    return BasicViaductFactory.create(
        tenantModuleInjectorFactory, List.of(DEFAULT_SCHEMA, EXTRAS_SCHEMA));
  }
}

File: core/x/remoteresolvers/starwars/main-server/src/main/kotlin/com/example/main/service/viaduct/ViaductConfiguration.kt
package com.example.main.service.viaduct

import io.micronaut.context.annotation.Bean
import io.micronaut.context.annotation.Factory
import jakarta.inject.Singleton
import viaduct.apiannotations.ExperimentalApi
import viaduct.engine.api.spi.ProxyResolverFactory
import viaduct.engine.runtime.tenantloading.ExecutionRegistryConfigSourceCollector
import viaduct.remote.config.RemoteResolverConfig
import viaduct.remote.config.RemoteResolverInitializer
import viaduct.remote.config.RemoteResolverSelection
import viaduct.service.SchemaScopeInfo
import viaduct.service.ViaductBuilder
import viaduct.service.api.Viaduct

const val DEFAULT_SCOPE_ID = "default"
const val EXTRAS_SCOPE_ID = "extras"
val DEFAULT_SCHEMA = SchemaScopeInfo.Scoped("publicSchema", setOf(DEFAULT_SCOPE_ID))
val EXTRAS_SCHEMA = SchemaScopeInfo.Scoped("publicSchemaWithExtras", setOf(DEFAULT_SCOPE_ID, EXTRAS_SCOPE_ID))

@Factory
class ViaductConfiguration(
    val tenantModuleInjectorFactory: MicronautTenantModuleInjectorFactory,
) {
    // Singleton so preDestroy targets the same instance the factory binds to.
    @Singleton
    @Bean(preDestroy = "close")
    fun remoteResolverInitializer(): RemoteResolverInitializer {
        val moduleConfigSources = ExecutionRegistryConfigSourceCollector.fromResources()

        // The sample always runs with the proxy enabled.
        return RemoteResolverInitializer(
            config = RemoteResolverConfig.fromEnvironment(enabled = true),
            selection =
                RemoteResolverSelection.fromModuleConfigSources(
                    selectedTenantNames = moduleConfigSources.map { it.tenantName }.toSet(),
                    moduleConfigSources = moduleConfigSources,
                ),
        )
    }

    @OptIn(ExperimentalApi::class)
    @Bean
    fun provideProxyResolverFactory(initializer: RemoteResolverInitializer): ProxyResolverFactory = initializer.initialize()

    @OptIn(ExperimentalApi::class)
    @Bean
    fun providesViaduct(proxyResolverFactory: ProxyResolverFactory): Viaduct =
        ViaductBuilder()
            .withScopedSchemas(listOf(DEFAULT_SCHEMA, EXTRAS_SCHEMA))
            .withTenantModuleInjectorFactory(tenantModuleInjectorFactory)
            .withProxyResolverFactory(proxyResolverFactory)
            .build()
}
